### DISCUSSION TITLE
Security system

### DISCUSSION THREAD

**Discussion by micahvdm** [HomeSpan/HomeSpan#300] [Q&A] [answered]
Hi all, I have successfully created a device with multiple motion sensors, but I’m wanting to implement the security system service. Has anyone done that with Homespan yet and if so, do you mind sharing how it works and how you implemented it?

**Comment by danilkorotkov** [comment]
https://github.com/HomeSpan/HomeSpan/blob/master/docs/ServiceList.md
![17E7B2EB-30D1-4D52-AB76-BE31FEDD27EF](https://user-images.githubusercontent.com/25537053/160290171-177cd64a-e0fe-4c54-9d41-92e26985c30e.jpeg)



[HAP-Specification-Non-Commercial-Version.pdf](https://github.com/HomeSpan/HomeSpan/files/8357994/HAP-Specification-Non-Commercial-Version.pdf)

**Reply by micahvdm** [reply]
Thanks, I’ve read through that, but what I’m not getting is how to implement it…does it require sub accessories or do you assign other sensors to it in home app?

**Reply by datjan** [reply]
Implementing the HomeKit security system is not easy.

There are two important characteristics. „target state“ and „current state“.
Possible states: Stay=0 / Away=1 / Night=2 / Disarmed=3 / Triggered=4

in default you can set both states to 1. and in case of triggered you set only current state to 4.
Everything hard to explain. Maybe you can check this sketch to learn how it works:
https://github.com/datjan/esp8266-homekit-securitysystem

**Comment by danilkorotkov** [comment]
HomeKit is protocol to connect your physical devices to your home. The alarm system is an independent external system with own logic: sensors, keys, alarm, users. What system do you want to create? Just call motion sensor panic to iPhone when armed ?

**Reply by micahvdm** [reply]
Yes exactly, but with iOS 15 motion sensor are only standard alerts. The security system allows critical alerts, which is what I’m actually after. Does this make sense?

**Reply by danilkorotkov** [reply]
In my case, I preferred to buy a ready-made solution
What sensors do you implemented? Wire connected?

**Reply by micahvdm** [reply]
Yeah, wired pir sensors. They work really well and are super cheap. HomeKit devices are not easy to get in South Africa and the ones that are are stupid expensive.

**Reply by danilkorotkov** [reply]
HomeKit-ready alarm  is expensive in most cases. But smart people created home assistant, home-bridge. Tuya isn’t hard to get. 
If you already realized pir sensor logic, why is difficult for you to create callback inside of  homespan?
Just poll your sensors in loop() of service and call setVal() to notify your Home

**Reply by danilkorotkov** [reply]
Or poll sensors in main loop() and notify through *ptr->SecuritySystemCurrentState->setVal(…)

**Reply by micahvdm** [reply]
I have the sensors working already. I think you might be misunderstanding. I have 5 motion sensors already created with homespan, but they do not allow critical alerts. That is why my question is about the security system service.

**Reply by danilkorotkov** [reply]
It doesn’t change what I said. Add virtual security service in your existing sketch and build some function to collect motion events and to call critical alerts if target state is armed

**Reply by danilkorotkov** [reply]
Create pointer to security service and use this pointer inside motion service or create callback in main and use it from sensor service to notify main then notify security if needed

**Reply by danilkorotkov** [reply]
Look at my recuperator sketch. RECUP *recuperator is pointer to get access homespan service and custom methods

**Comment by datjan** [comment]
I've made an example for the security system using HomeSpan:

https://github.com/datjan/esp32-homekit-securitysystem

**Comment by micahvdm** [comment]
So I managed to get this all set up, but the problem now is that I need to add a button in home or custom characteristic to be able to trigger the alarm to state 4 in automations. I was using homebridge security system and this exposes a switch called siren that basically sets the alarm to state 4 “triggered”. Is this possible to do in HomeSpan?

**Comment by HomeSpan** [comment]
There are lots of different ways of doing this.  One possibility would be to add another Accessory with a simple Switch Service containing the On Characteristic (see the HAP-R2 doc for detail) to your sketch.  This will show up as a generic switch in your Home App.  Alternatively, you can add a Switch Service to the same Accessory that contains your Security System Service.  In this case the switch will show up next to the Home/Away/Disarm selector on the Alarm page.  The Home App allows you to show services within the same accessory as separate tiles, so this will allow you to experiment with different placement of the switch.

Then add code to the update() potion of your Switch Service (in your HomeSpan sketch) that resets the Security System Current State to "4" when the On Characteristic is set to 1.  Since the Security System Current State is a Characteristic of Security System Service, not the Switch Service, you'll need to save the pointer to the Security Service when you instantiate it so that you can pass that pointer (or make it a global variable) to the Switch Service.

**Comment by HomeSpan** [comment, marked as answer]
Here's a complete sketch that illustrates the idea above:

```C++
#include "HomeSpan.h"

/////////////////////////////////

struct SecuritySystem : Service::SecuritySystem{

  Characteristic::SecuritySystemCurrentState current{3};
  Characteristic::SecuritySystemTargetState target{3};
  
  SecuritySystem() : Service::SecuritySystem(){
     target.setValidValues(3,0,1,3);                // only allow Stay Arm, Away Arm, and Disarm (exclude Night Arm)
   }

  boolean update() override{
    current.setVal(target.getNewVal());
    return(true);
  }
  
};

/////////////////////////////////

SecuritySystem *securitySystem;         // create a global pointer to a SecuritySystem structure

/////////////////////////////////

struct Siren : Service::Switch{

  Characteristic::On sirenState{0};

  Siren() : Service::Switch(){
    new Characteristic::Name("Siren");  
  }

  boolean update() override{
    if(sirenState.getNewVal()){              // if siren switch is activated
      securitySystem->current.setVal(4);     // trigger the alarm
      Serial.printf("Siren Activated\n");
    } else {
      securitySystem->current.setVal(securitySystem->target.getVal());     // reset current state to original target state
      Serial.printf("Siren Reset\n");      
    }
    return(true);
  }
  
};

/////////////////////////////////

void setup() {
  
  Serial.begin(115200);

  homeSpan.begin(Category::SecuritySystems,"HomeSpan Alarm");

  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Identify();
    securitySystem = new SecuritySystem();      // note we save the pointer so it can be referenced by Siren (in code above)
    new Siren();    
      
} // end of setup()

//////////////////////////////////////

void loop(){
  
  homeSpan.poll();
  
} // end of loop()
```

**Reply by micahvdm** [reply]
Amazing! Thanks for the reply, I'm gonna add this now and test! Such a great library!!

**Reply by adamlilley88** [reply]
did you get anywhere with this? I am attenmpting the same thing, I have my sensors configured with homespan, and showing up in homekit. but next step in the incorporate the security system natively on the esp32
